api: Deprecate LoadBalancer.handleResolvedAddresses()

SreeramdasLavanya opened the pull request:
Also deprecate its companion canHandleEmptyAddressListFromNameResolution().
Also fixup the Javadoc to align with the arguments/return values, so that people
would have a better idea of how to use it.

Fixes #11194

SreeramdasLavanya committed:
SynchronizationContextTest changes for scheduleFixedDelay with Duration

SreeramdasLavanya committed:
Revert "SynchronizationContextTest changes for scheduleFixedDelay with Duration"

This reverts commit ab9704560525468d28294c7c6c0e4266e79efab4.

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
api: Javadoc changes for LoadBalancer method signatures

SreeramdasLavanya commented:
Below points have been addressed:

javadoc for public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses)

- refers to non-existent argument servers

javadoc for public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses)

- refers to an argument named addresses
- refers to returning values not of the actual return type

kannanjgithub commented on api/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -179,12 +179,12 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
    * single list if needed.
    *
-   * <p>Implementations can choose to reject the given addresses by returning {@code false}.
+   * <p>Implementations can choose to reject the given addresses by returning {@code UNAVAILABLE}.
```
Replace with {@code Status.UNAVAILABLE} . Also below.

SreeramdasLavanya committed:
api: Changes for UNAVAILABLE to Status.UNAVAILABLE and Status.OK

SreeramdasLavanya committed:
api: Check style fix for UNAVAILABLE to Status.UNAVAILABLE and Status.OK

SreeramdasLavanya committed:
api: Check style fix for Status.UNAVAILABLE and Status.OK

SreeramdasLavanya committed:
api: Check style fix for Status.UNAVAILABLE

SreeramdasLavanya committed:
api: Check style fix for Status.UNAVAILABLE

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
api/util/services: Deprected handleResolvedAddresses method and related changes

SreeramdasLavanya committed:
interop-testing: Changes for deprecated handleResolvedAddresses method and related changes

SreeramdasLavanya committed:
xds: Changes for deprecated handleResolvedAddresses method and related changes

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
api: Reverted changes for LoadBalancer and LoadBalancerTest

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
util: Replaced handleResolvedAddress to acceptResolvedAddresses.

SreeramdasLavanya committed:
api: Reverted back to acceptResolvedAddress

SreeramdasLavanya commented on api/src/test/java/io/grpc/LoadBalancerTest.java:
```diff
@@ -240,9 +240,8 @@ public void handleResolvedAddresses_delegatesToAcceptResolvedAddresses() {
         @Override
-        public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
           resultCapture.set(resolvedAddresses);
```
Rolled back please review

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

shivaspeaks commented on api/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -156,15 +156,16 @@ public String toString() {
    * @param resolvedAddresses the resolved server addresses, attributes, and config.
    * @since 1.21.0
    */
+  @Deprecated
```
Add a javadoc on Deprecated and what to use instead.

ejona86 commented on util/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java:
```diff
@@ -112,20 +112,6 @@ public GracefulSwitchLoadBalancer(Helper helper) {
   }
 
-  @Override
-  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
```
You can't just delete this. It is needed as long as this method in on the API. Mark it `@Deprecated`; `@Deprecated` functions can all other deprecated functions without warning.

SreeramdasLavanya replied:
Marked as ```@Deprecated```

ejona86 commented on util/src/main/java/io/grpc/util/MultiChildLoadBalancer.java:
```diff
@@ -193,7 +193,7 @@ private List<ChildLbState> updateChildrenWithResolvedAddresses(
       if (entry.getValue() != null) {
         childLbState.setResolvedAddresses(entry.getValue()); // update child
-        childLbState.lb.handleResolvedAddresses(entry.getValue()); // update child LB
+        childLbState.lb.acceptResolvedAddresses(entry.getValue()); // update child LB
```
This class needs to fix its handling of the `Status`. I think we should leave it as-is and create a tracking bug for it.

SreeramdasLavanya replied:
Leaving it as-is(handleResolvedAddresses) is resulting in build failure because of the deprecated warning, so i'm retaining this acceptResolvedAddresses for now to proceed further. Please confirm if this is ok and handling the ```Status``` can be done as part of a new defect.

shivaspeaks replied:
Hmmm, we can definitely suppress warning here and create a tracking bug. Add a comment with TODO mentioning the tracking bug.

ejona86 replied:
The tracking bug added was the already-existing bug this PR was fixing, which means it was going to be closed when this PR went in thus not be a tracking bug for the problem. But https://github.com/grpc/grpc-java/pull/11894 is now in, so this doesn't matter any more.

SreeramdasLavanya replied:
Merged with master branch latest changes

ejona86 commented on util/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java:
```diff
@@ -171,7 +171,7 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     }
 
-    switchLb.handleResolvedAddresses(
+    switchLb.acceptResolvedAddresses(
```
return the status from this call

ejona86 commented on util/src/test/java/io/grpc/util/GracefulSwitchLoadBalancerTest.java:
```diff
@@ -125,17 +125,17 @@ public void switchTo_handleResolvedAddressesAndNameResolutionErrorForwardedToLat
     ResolvedAddresses addresses = newFakeAddresses();
-    gracefulSwitchLb.handleResolvedAddresses(addresses);
-    verify(lb0).handleResolvedAddresses(addresses);
+    gracefulSwitchLb.acceptResolvedAddresses(addresses);
```
Don't change any of these. These shouldn't have caused warnings because they were already marked `@Deprecated`. They are testing the old code flow and need to remain until it is deleted. When the old flow is deleted, the entire set of "// OLD TESTS" will be outright deleted, as we already have tests below that test acceptResolvedAddresses (and test it more fully, as it checks the return value).

SreeramdasLavanya replied:
Handled as mentioned

ejona86 commented on util/src/main/java/io/grpc/util/ForwardingLoadBalancer.java:
```diff
@@ -30,8 +30,8 @@ public abstract class ForwardingLoadBalancer extends LoadBalancer {
   protected abstract LoadBalancer delegate();
 
   @Override
-  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
```
This will need some more thought. This change would prevent extending classes from seeing the addresses, if they are only overriding handleResolvedAddresses().

SreeramdasLavanya replied:
Can we mark this existing method as deprecated and introduce an override for acceptResolvedAddresses, please confirm

ejona86 replied:
That would still break an implementation only implementing `handleResolvedAddresses()`

shivaspeaks replied:
Marking `handleResolveAddresses()` deprecated and also calling newly introduced `acceptResolvedAddresses()` within it should we the way to go forward. Because acceptResolvedAddresses will delegate it to LoadBalancer and there we are using recursionCount to deal with things. So it works as far as I can think.

Something like,

```
  @Deprecated
  @Override
  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
      acceptResolvedAddresses(resolvedAddresses);
  }

  @Override
  public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
      return delegate().acceptResolvedAddresses(resolvedAddresses);
  }
```

SreeramdasLavanya replied:
Handled as mentioned

SreeramdasLavanya replied:
@shivaspeaks Attempting to use the @InlineMe annotation did not resolve the error as it mandates the method to be marked as final which is unsuitable, suppressing all warnings also failed to address the issue. Anything you would like to suggest here.

ejona86 replied:
Revert this method's implementation back to forwarding to `handleResolvedAddresses()`. Everything (in our code base) extending this class has both methods implemented. You can then remove the SuppressWarnings.

ejona86 commented on api/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -156,15 +156,16 @@ public String toString() {
+   * into a single list if needed.
    *
-   * <p>Implementations should not modify the given {@code servers}.
+   * <p>Implementations should not modify the given {@code resolvedAddresses}.
```
Delete this line. `servers` had been a collection. `ResolvedAddresses` doesn't let itself be modified.

SreeramdasLavanya replied:
Deleted the same

ejona86 commented on api/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -156,15 +156,16 @@ public String toString() {
-   * {@code servers} contained in {@link EquivalentAddressGroup} should be considered equivalent
-   * but may be flattened into a single list if needed.
+   * Handles newly resolved addresses and metadata attributes from name resolution system.
+   * {@link EquivalentAddressGroup} addresses should be considered equivalent but may be flattened
```
"Addresses in {@link EquivalentAddressGroup}"? It isn't clear what EAG addresses is talking about; e.g., are EAG addresses a type of address?

SreeramdasLavanya replied:
@code servers is removed from the existing javadoc. Can you elaborate what is expected here?

shivaspeaks replied:
Use "Addresses in {@link EquivalentAddressGroup}" instead of "{@link EquivalentAddressGroup}"

shivaspeaks replied:
Use "Addresses in {@link EquivalentAddressGroup}" instead of "{@link EquivalentAddressGroup} addresses"!

SreeramdasLavanya replied:
Addressed as mentioned

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'master' into FixIssue-11194

# Conflicts:
#	util/src/main/java/io/grpc/util/MultiChildLoadBalancer.java

SreeramdasLavanya committed:
api/util: Addressed the re-review points

SreeramdasLavanya committed:
api/util: Resolved Deprecation warnings

SreeramdasLavanya committed:
api/util: Resolved checkstyle issues

SreeramdasLavanya committed:
api/util: Resolved deprecation warnings

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
util/xds: Resolved deprecation warnings

SreeramdasLavanya committed:
util: Resolved deprecation warnings

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
util: Included suppress warnings for updateChildrenWithResolvedAddresses method

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

ejona86 commented on util/src/main/java/io/grpc/util/MultiChildLoadBalancer.java:
```diff
@@ -191,6 +192,7 @@ private List<ChildLbState> updateChildrenWithResolvedAddresses(
       newChildLbStates.add(childLbState);
       if (entry.getValue() != null) {
+        //TODO - https://github.com/grpc/grpc-java/issues/11194
         childLbState.lb.handleResolvedAddresses(entry.getValue()); // update child LB
```
https://github.com/grpc/grpc-java/pull/11894 swapped this to acceptResolvedAddresses()

SreeramdasLavanya replied:
Merged with master branch latest changes

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'master' into FixIssue-11194

# Conflicts:
#	util/src/main/java/io/grpc/util/MultiChildLoadBalancer.java

SreeramdasLavanya committed:
util: Changed deprecated method to newly added method

SreeramdasLavanya committed:
util: Removed delegate for acceptResolved method

SreeramdasLavanya committed:
util: Removed override annotation

SreeramdasLavanya committed:
Merge branch 'grpc:master' into master

SreeramdasLavanya committed:
Merge branch 'master' into FixIssue-11194

# Conflicts:
#	services/src/main/java/io/grpc/protobuf/services/HealthCheckingLoadBalancerFactory.java

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
util: included surpresswarning annotation to fix Inlineable issue

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

SreeramdasLavanya committed:
Merge branch 'grpc:master' into FixIssue-11194

ejona86 commented on xds/src/test/java/io/grpc/xds/PriorityLoadBalancerTest.java:
```diff
@@ -145,6 +145,7 @@ public void tearDown() {
     assertThat(fakeClock.getPendingTasks()).isEmpty();
   }
 
+  @Deprecated
```
Revert these changes? The tests were converted to acceptResolvedAddresses.

ejona86 commented on xds/src/test/java/io/grpc/xds/WrrLocalityLoadBalancerTest.java:
```diff
@@ -107,6 +107,7 @@ public void setUp() {
     loadBalancer = new WrrLocalityLoadBalancer(mockHelper, lbRegistry);
   }
 
+  @Deprecated
```
Revert these changes? The tests were converted to acceptResolvedAddresses.

ejona86 commented on util/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java:
```diff
@@ -49,6 +49,13 @@
+     * @deprecated  As of release 1.69.0,
+     *     use instead {@link #acceptResolvedAddresses(ResolvedAddresses)}
+     */
+    @Deprecated
```
This method is not deprecated, and should not be marked deprecated.

ejona86 commented on xds/src/test/java/io/grpc/xds/WeightedTargetLoadBalancerTest.java:
```diff
@@ -179,6 +179,7 @@ public void tearDown() {
     }
   }
 
+  @Deprecated
```
We want to fix these. I've sent out https://github.com/grpc/grpc-java/pull/12053

ejona86 commented on util/src/test/java/io/grpc/util/ForwardingLoadBalancerTest.java:
```diff
@@ -45,6 +45,8 @@ public void allMethodsForwarded() throws Exception {
         new TestBalancer(),
         Arrays.asList(
-            LoadBalancer.class.getMethod("acceptResolvedAddresses", ResolvedAddresses.class)));
+            LoadBalancer.class.getMethod("acceptResolvedAddresses", ResolvedAddresses.class),
```
acceptResolvedAddresses should be removed already with the current code. After you change handleResolverAddresses to forward, it should be removed as well.

shivaspeaks committed:
Merge branch 'master' of https://github.com/grpc/grpc-java into FixIssue-11194

shivaspeaks committed:
address comments

shivaspeaks committed:
use acceptResolvedAddresses in ClusterImplLoadBalancer

ejona86 committed:
Remove useless docs added to GracefluwSwitchLb

ejona86 committed:
Mark forwarding implementation deprecated instead of deleting

We want to keep the old behavior until the old usages are removed.

ejona86 committed:
Delete random formatting change to file

ejona86 committed:
Tweak doc tweaks

ejona86 committed:
Don't implement ForwardingLb.acceptResolvedAddresses()

Any old implementations that still implement handleResolvedAddresses()
will have their implementationed "skipped" by new callers of
acceptResolvedAddresses(). We've been having such forwarders implement
both methods instead. Since there's very few usages of
handleResolvedAddresses() left, out-of-tree implementations should
generally be fine just swapping to acceptResolvedAddresses.

ejona86 committed:
Fix missed doc mentioning UNAVAILABLE

ejona86 committed:
Restore removed Override annotation

ejona86 committed:
Remove useless doc

ejona86 committed:
The test was for the old implementation

It should have been completely deleted if the method was being deleted.
In this case, since I've restored the method we just need Deprecated on
the test.

ejona86 reviewed (approved):
@shivaspeaks, you can see now how the change is basically _just_ the deprecation annotation and the one ClusterImplLoadBalancer fix. My best guess for why ClusterImplLoadBalancer wasn't already migrated is because we might have been in the middle of L74 and didn't want to do any unnecessary changes that would conflict.

shivaspeaks committed:
deprecate canHandleEmptyAddressListFromNameResolution

shivaspeaks committed:
Merge branch 'FixIssue-11194' of https://github.com/SreeramdasLavanya/grpc-java into FixIssue-11194

ejona86 committed:
Fix style; more precise language
